### ISSUE TITLE
Deserializing ViewId can panic

### ISSUE THREAD

**Issue by portstrom** [xi-editor/xi-editor#832]
`ViewId` has a custom deserializer implementation that [panics when receiving a string that doesn't match the expected format](https://github.com/xi-editor/xi-editor/blob/master/rust/core-lib/src/tabs.rs#L856). Deserializers should never panic, but return an error. **Suggested solution:** Remove custom serializer and deseralizer, as they are pointless.

**Comment by cmyr** [comment]
This is fair, the deserializer should return an error. For the second comment, I find the phrasing a little unhelpful; perhaps by "they are pointless", you've intended something like ["I do not see the point of them."](https://en.wikipedia.org/wiki/Wikipedia:Chesterton%27s_fence) ?

In any case sending malformed RPCs is currently not something we recover from; and view identifiers, (for the purpose of RPC), are strings, and they all originate in the core. Clients should only use identifiers that they have received from core.

**Comment by portstrom** [comment]
I prefer [Occam's razor](https://en.wikipedia.org/wiki/Occam%27s_razor) to Chesterton's fence. :wink: Anyway, I think it's the responsibility of anyone who makes the program more complex in a seemingly pointless way to document the motivation. Otherwise people will spend work patching the unnecessarily complex code instead of simplifying it, being afraid that anything that isn't already as simple as possible may have an esoteric reason not to be simple. It would be interesting to know why it's a good idea to have extra code to serialize to a format that is less efficient on the wire and requires clients to treat view IDs as the less efficient `Rc<String>` when `usize` would have been fine.

**Comment by cmyr** [comment]
In terms of understanding project history: this is always a fine line. In this case I think that `git blame` does a reasonable job of explaining how we got to this point.

To skip ahead to the rationale: string identifiers on the wire make RPCs easier to debug, and allows us to more easily to visually distinguish a view identifier from a buffer or plugin identifier in logs.

Originally, these identifiers were also strings in core; at some point we moved to using a newtype struct for this, and then later as we built out some of the early plugin stuff and started having to handle crossing more thread barriers, it made sense to switch to using `Copy` types internally. We could have chosen to change the protocol to just use a numerical type at this point, but we would have been sacrificing readability, and we would have broken all existing clients.

It makes sense to me that at some point we refactor and have a unified `xi-protocol` crate that can be used from both the frontend and the backend; this would let _Rust_ frontends take advantage of the better internal `ViewId` representation, and would mean less work is duplicated. To date, however, nobody has taken on this project. The `ViewId` type _is_ public, so a rust frontend that includes the `xi-core` crate can use it, and get the usize representation.
